src/terrain/clients/iplant_groups.clj

```diff
@@ -166,3 +174,55 @@
         group  (format "%s:%s" folder name)]
     (verify-group-exists client user group)
     (c/remove-group-members client user group members)))
+
+;; Team Functions
+
+(def ^:private team-group-type "group")
+
+(defn- get-team-folder-name [client & [user]]
+  (if-not user
+    (c/build-folder-name client "teams")
+    (c/build-folder-name client (format "teams:%s" user))))
+
+(defn- ensure-team-folder-exists [client user]
+  (ensure-folder-exists client (config/grouper-user) (get-team-folder-name client))
+  (ensure-folder-exists client user (get-team-folder-name client user)))
+
+(defn- get-teams* [client user search-folder lookup-fn]
+  (let [folder (get-team-folder-name client)]
+    (get-groups* search-folder (partial format-group folder) client user lookup-fn)))
+
+(defn- filter-teams [search result]
+  (update-in result [:groups] (partial filter (comp (partial re-find (re-pattern (str "\\Q" search))) :name))))
+
+(defn- find-groups-with-member [client user member search-folder search]
+  (let [result (c/list-subject-groups client user member search-folder)]
+    (if search
+      (filter-teams search result)
+      result)))
+
+;; This function kind of uses a hack. A search string is required, but if we make it the
+;; same as the folder name then that approximates listing all groups in the folder. An
+;; update to iplant-groups will be required to eliminate this hack.
+(defn get-teams [user {:keys [search creator member]}]
+  (let [client (get-client)
+        folder (get-team-folder-name client creator)]
+    (->> (cond member (fn [_] (find-groups-with-member client user member folder search))
+               search (partial c/find-groups client user search)
+               :else  (partial c/find-groups client user))
+         (get-teams* client user folder))))
+
+(defn add-team [user {:keys [name description public_privileges]}]
+  (let [client (get-client)
+        folder (get-team-folder-name client user)]
+    (ensure-team-folder-exists client user)
+    (let [full-name (str folder ":" name)
+          group     (c/add-group client user full-name team-group-type description)]
+      (dorun (map (partial c/grant-group-privilege client user full-name c/public-user) public_privileges))
```

Comment:
Yeah, `doseq` would have worked well here. My brain usually goes to `dorun map` instead for some reason. I may change this the next time I refactor the code.